Mittons.Fixtures.Tests.Unit/Docker/Containers/HttpContainerTests.cs

```diff
@@ -0,0 +1,177 @@
+            _containers.Add(container);
+
+            // Act
+            await container.InitializeAsync();
+
+            // Assert
+            Assert.Equal(new Uri($"https://{expectedHost}:{expectedPort}"), container.SecureHttpClient.BaseAddress);
+        }
+
+        [Theory]
+        [InlineData(49621, "192.168.0.1")]
+        [InlineData(49622, "192.168.0.2")]
+        public async Task InitializeAsync_WhenTheUnsecurePortIsNotBound_ExpectHttpClientToHaveBaseAddressSetToTheDefault(int hostPort, string containerIpAddress)
+        {
+            // Arrange
+            var containerPort = 80;
+
+            var gatewayMock = new Mock<IDockerGateway>();
+            gatewayMock.Setup(x => x.ContainerGetDefaultNetworkIpAddressAsync(It.IsAny<string>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(IPAddress.Parse(containerIpAddress));
+            gatewayMock.Setup(x => x.ContainerGetHealthStatusAsync(It.IsAny<string>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(HealthStatus.Healthy);
+            gatewayMock.Setup(x => x.ContainerGetHostPortMappingAsync(It.IsAny<string>(), "tcp", containerPort, CancellationToken.None)).ReturnsAsync(hostPort);
+
+            var expectedHost = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "localhost" : containerIpAddress;
+            var expectedPort = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? hostPort : containerPort;
+
+            var container = new HttpContainer(gatewayMock.Object, Guid.NewGuid(), new Attribute[] { new Image("mendhak/http-https-echo") });
+            _containers.Add(container);
+
+            // Act
+            await container.InitializeAsync();
+
+            // Assert
+            Assert.Equal(new Uri($"http://{expectedHost}:{expectedPort}"), container.UnsecureHttpClient.BaseAddress);
+        }
+
+        [Theory]
+        [InlineData(49621, "192.168.0.1")]
+        [InlineData(49622, "192.168.0.2")]
+        public async Task InitializeAsync_WhenTheSecurePortIsNotBound_ExpectHttpClientToHaveBaseAddressSetToTheDefault(int hostPort, string containerIpAddress)
+        {
+            // Arrange
+            var containerPort = 443;
+
+            var gatewayMock = new Mock<IDockerGateway>();
+            gatewayMock.Setup(x => x.ContainerGetDefaultNetworkIpAddressAsync(It.IsAny<string>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(IPAddress.Parse(containerIpAddress));
+            gatewayMock.Setup(x => x.ContainerGetHealthStatusAsync(It.IsAny<string>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(HealthStatus.Healthy);
+            gatewayMock.Setup(x => x.ContainerGetHostPortMappingAsync(It.IsAny<string>(), "tcp", containerPort, CancellationToken.None)).ReturnsAsync(hostPort);
+
+            var expectedHost = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "localhost" : containerIpAddress;
```

Comment:
So when it comes to tests, I'm less concerned about DRY since that can unintentionally lead to obfuscated tests, or starting to build tests that are tightly coupled to a shared resource and become difficult to maintain.  Not saying I'm always against it, just pointing out that I think tests are an area where normal best practices aren't always best to adhere to.

Rather than create some function all these tests run, I think I'd rather figure out how to run tests in github against a windows instance, and create windows and linux versions of the tests, so we don't need these ternaries.

Reply:
That's a neat idea. I wasn't necessarily suggesting a function that get called per test, but wondering if there was a way to reduce the need to know in general. Your idea of specifying the OS environment for the tests to run in pairs nicely with being able to remove this ternary from this area somehow.
Often the way we naturally flow towards writing tests can show us a new angle of the design which leads us to making more extensible and maintainable systems.